usim/_basics/resource.py

```diff
@@ -0,0 +1,208 @@
+        new_levels = self._available.value - amounts
+        await self._available.set(new_levels)
+
+    def borrow(self, **amounts: T) -> 'BorrowedResources[T]':
+        """
+        Temporarily borrow resources for a given context
+
+        :param amounts: resource levels to borrow
+        :return: async context to borrow resources
+        """
+        borrowed_levels = self._levels_type(**amounts)
+        assert self._levels_type.zero <= borrowed_levels,\
+            'cannot borrow negative amounts'
+        return BorrowedResources(self, borrowed_levels)
+
+
+class BorrowedResources(BaseResources[T]):
+    """
+    Fixed supply of named resources temporarily taken from another resource supply
+    """
+    @property
+    def _levels_type(self):
+        return self._resources._levels_type
+
+    def __init__(self, resources: 'BaseResources', capacity: ResourceLevels):
+        self._resources = resources
+        self._capacity = capacity
+        self._available = Tracked(self._levels_type.zero)
+
+    async def __aenter__(self):
+        await (self._resources._available >= self._capacity)
+        await self._resources.__remove_resources__(self._capacity)
+        await self.__insert_resources__(self._capacity)
+        return self
+
+    async def __aexit__(self, exc_type, exc_val, exc_tb):
+        if exc_type is GeneratorExit:
+            # we are killed forcefully and cannot perform async operations
+            # dispatch a new activity to release our resources eventually
+            __LOOP_STATE__.LOOP.schedule(
+                self._resources.__insert_resources__(self._capacity)
+            )
+            __LOOP_STATE__.LOOP.schedule(
+                self.__remove_resources__(self._capacity)
+            )
+        else:
+            await self._resources.__insert_resources__(self._capacity)
+            await self.__remove_resources__(self._capacity)
+            # TODO: forcefully kill off anyone holding our resources?
+
+    def borrow(self, **amounts: T) -> 'BorrowedResources[T]':
+        borrowing = super().borrow(**amounts)
+        assert self._capacity >= borrowing._capacity,\
+            'cannot borrow beyond capacity'
+        return borrowing
+
+
+class Capacities(BorrowedResources[T]):
+    r"""
+    Fixed supply of named resources which can be temporarily borrowed
+
+    The resources and their maximum capacity are defined
+    when the resource supply is created.
+    Afterwards, it is only possible to temporarily :py:meth:`borrow`
+    resources:
+
+    .. code:: python3
+
+        # create a limited supply of resources
+        resources = Capacities(cores=8, memory=16000)
+
+        # temporarily remove resources
+        async with resources.borrow(cores=2, money=4000):
+            await computation
+
+    A :py:class:`~.Capacities` guarantees that its resources are conserved and
+    cannot be leaked. Once resources are :py:meth:`~.borrow`\ ed, they can
+    always be returned promptly.
+    """
```

Comment:
I do not really see the difference between `Capacities` and `BorrowedResources` given the class description.

Reply:
The only difference is that ``BorrowedResources`` are borrowed from *another* resource, while ``Capacities`` are constructed alone. We could construct ``Capacities`` as ``Resources().borrow`` (which is basically what ``Capacities`` does under the hood).

Do you think it would help to point that out explicitly?